### PR TITLE
[Feature]Parallel execution of compaction tasks under one tablet

### PR DESCRIPTION
**Pull request by choubenson**
**Background**
Small rowsets accumulation may affect writing and querying. If a time-consuming compaction task is being executed under a tablet, user write traffic increases at the same time, and a large number of small rowsets will be generated in a short period of time, which can not be compacted in time. After reaching 1000, the write would be blocked, and It will also seriously affect query performance.

**Solution**
To solve the problem above, this PR implements the parallel execution of compaction tasks under a tablet, and currently only supports the new event-triggered compaction framework and sizeTiered file selection strategy. 
The specific implementation method is as follows: 
- Add a bool status `is_compacting` to each rowset, which means "whether it is being compacted" 
- When the event triggers the tablet to select files, regardless of whether the tablet is executing a compaction task, the file selection will be executed, and the rowset whose `is_compacting` is true will be filtered. 
- When the compaction task starts, the `is_compacting` of the rowsets will be set to true, and when it ends, `is_compacting` will be set to false.

See the first solution in the design doc [document](https://apache-iotdb.feishu.cn/docx/PPv8d17BQoRIKIxJ4TEcfayInuh?from=from_copylink) for details.

## What type of PR is this:

- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [x] Yes, this PR will result in a change in behavior.
- [ ] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [x] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 73 / 146 (50.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/storage/compaction_manager.h|0|5|00.00%|[69, 70, 71, 73, 83]|
|:large_blue_circle:|src/storage/tablet.cpp|13|68|19.12%|[1178, 1179, 1180, 1181, 1182, 1184, 1185, 1186, 1188, 1189, 1190, 1191, 1192, 1194, 1195, 1196, 1197, 1198, 1199, 1202, 1205, 1206, 1207, 1209, 1210, 1211, 1212, 1214, 1215, 1216, 1218, 1219, 1220, 1222, 1223, 1224, 1226, 1227, 1228, 1229, 1230, 1232, 1233, 1234, 1235, 1237, 1238, 1239, 1240, 1241, 1243, 1245, 1247, 1551, 1552]|
|:large_blue_circle:|src/storage/rowset/rowset.h|1|2|50.00%|[290]|
|:large_blue_circle:|src/storage/tablet.h|3|6|50.00%|[177, 182, 210]|
|:large_blue_circle:|src/storage/size_tiered_compaction_policy.cpp|10|14|71.43%|[213, 215, 216, 217]|
|:large_blue_circle:|src/storage/compaction_manager.cpp|42|47|89.36%|[413, 414, 416, 516, 517]|
|:large_blue_circle:|src/storage/compaction_task.h|2|2|100.00%|[]|
|:large_blue_circle:|src/storage/compaction_task.cpp|2|2|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 75 / 146 (51.37%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/storage/compaction_manager.h|0|5|00.00%|[69, 70, 71, 73, 83]|
|:large_blue_circle:|src/storage/tablet.cpp|15|68|22.06%|[1178, 1179, 1180, 1181, 1182, 1184, 1185, 1186, 1188, 1189, 1190, 1191, 1192, 1194, 1195, 1196, 1197, 1198, 1199, 1202, 1205, 1206, 1207, 1209, 1210, 1211, 1212, 1214, 1215, 1216, 1218, 1219, 1220, 1222, 1223, 1224, 1226, 1227, 1228, 1229, 1230, 1232, 1233, 1234, 1235, 1237, 1238, 1239, 1240, 1241, 1243, 1245, 1247]|
|:large_blue_circle:|src/storage/rowset/rowset.h|1|2|50.00%|[290]|
|:large_blue_circle:|src/storage/tablet.h|3|6|50.00%|[177, 182, 210]|
|:large_blue_circle:|src/storage/size_tiered_compaction_policy.cpp|10|14|71.43%|[213, 215, 216, 217]|
|:large_blue_circle:|src/storage/compaction_manager.cpp|42|47|89.36%|[413, 414, 416, 516, 517]|
|:large_blue_circle:|src/storage/compaction_task.h|2|2|100.00%|[]|
|:large_blue_circle:|src/storage/compaction_task.cpp|2|2|100.00%|[]|

File: be/src/storage/compaction_manager.h [hunk 1]
```diff
@@ -66,8 +66,11 @@ class CompactionManager {
     void get_running_status(std::string* json_result);
 
     uint16_t running_tasks_num() {
```
[the remark this answers is not in the document]
Comment (choubenson) [reply be/src/storage/compaction_manager.h:68]: Resolved.

File: .gitignore [hunk 2]
```diff
@@ -33,6 +34,10 @@ fe/ut_ports
 fe/*/target
 fe/fe-core/gen
 fe/fe-common/.classpath
+fe/meta
```
[the remark this answers is not in the document]
Comment (choubenson) [reply .gitignore]: Resolved.

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 75 / 150 (50.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/storage/compaction_manager.h|0|4|00.00%|[84, 85, 86, 88]|
|:large_blue_circle:|src/storage/tablet.cpp|15|71|21.13%|[1178, 1179, 1180, 1181, 1182, 1184, 1185, 1186, 1188, 1189, 1190, 1191, 1192, 1194, 1195, 1196, 1197, 1198, 1199, 1200, 1201, 1204, 1207, 1208, 1209, 1211, 1212, 1213, 1214, 1216, 1217, 1218, 1220, 1221, 1222, 1224, 1225, 1226, 1228, 1229, 1230, 1231, 1232, 1234, 1235, 1236, 1237, 1239, 1240, 1241, 1242, 1243, 1245, 1247, 1248, 1250]|
|:large_blue_circle:|src/storage/rowset/rowset.h|1|2|50.00%|[290]|
|:large_blue_circle:|src/storage/tablet.h|3|6|50.00%|[177, 182, 210]|
|:large_blue_circle:|src/storage/size_tiered_compaction_policy.cpp|10|14|71.43%|[213, 215, 216, 217]|
|:large_blue_circle:|src/storage/compaction_manager.cpp|42|49|85.71%|[413, 414, 415, 416, 417, 517, 518]|
|:large_blue_circle:|src/storage/compaction_task.h|2|2|100.00%|[]|
|:large_blue_circle:|src/storage/compaction_task.cpp|2|2|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 75 / 150 (50.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/storage/compaction_manager.h|0|4|00.00%|[84, 85, 86, 88]|
|:large_blue_circle:|src/storage/tablet.cpp|15|71|21.13%|[1178, 1179, 1180, 1181, 1182, 1184, 1185, 1186, 1188, 1189, 1190, 1191, 1192, 1194, 1195, 1196, 1197, 1198, 1199, 1200, 1201, 1204, 1207, 1208, 1209, 1211, 1212, 1213, 1214, 1216, 1217, 1218, 1220, 1221, 1222, 1224, 1225, 1226, 1228, 1229, 1230, 1231, 1232, 1234, 1235, 1236, 1237, 1239, 1240, 1241, 1242, 1243, 1245, 1247, 1248, 1250]|
|:large_blue_circle:|src/storage/rowset/rowset.h|1|2|50.00%|[290]|
|:large_blue_circle:|src/storage/tablet.h|3|6|50.00%|[177, 182, 210]|
|:large_blue_circle:|src/storage/size_tiered_compaction_policy.cpp|10|14|71.43%|[213, 215, 216, 217]|
|:large_blue_circle:|src/storage/compaction_manager.cpp|42|49|85.71%|[413, 414, 415, 416, 417, 517, 518]|
|:large_blue_circle:|src/storage/compaction_task.h|2|2|100.00%|[]|
|:large_blue_circle:|src/storage/compaction_task.cpp|2|2|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 75 / 150 (50.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/storage/compaction_manager.h|0|4|00.00%|[84, 85, 86, 88]|
|:large_blue_circle:|src/storage/tablet.cpp|15|71|21.13%|[1192, 1193, 1194, 1195, 1196, 1198, 1199, 1200, 1202, 1203, 1204, 1205, 1206, 1208, 1209, 1210, 1211, 1212, 1213, 1214, 1215, 1218, 1221, 1222, 1223, 1225, 1226, 1227, 1228, 1230, 1231, 1232, 1234, 1235, 1236, 1238, 1239, 1240, 1242, 1243, 1244, 1245, 1246, 1248, 1249, 1250, 1251, 1253, 1254, 1255, 1256, 1257, 1259, 1261, 1262, 1264]|
|:large_blue_circle:|src/storage/rowset/rowset.h|1|2|50.00%|[292]|
|:large_blue_circle:|src/storage/tablet.h|3|6|50.00%|[177, 182, 210]|
|:large_blue_circle:|src/storage/size_tiered_compaction_policy.cpp|10|14|71.43%|[213, 215, 216, 217]|
|:large_blue_circle:|src/storage/compaction_manager.cpp|42|49|85.71%|[413, 414, 415, 416, 417, 517, 518]|
|:large_blue_circle:|src/storage/compaction_task.h|2|2|100.00%|[]|
|:large_blue_circle:|src/storage/compaction_task.cpp|2|2|100.00%|[]|

**Review by meegoo — APPROVED** [review]

File: be/src/storage/compaction_manager.cpp [hunk 3]
```diff
@@ -338,8 +332,17 @@ bool CompactionManager::register_task(CompactionTask* compaction_task) {
     std::lock_guard lg(_tasks_mutex);
     TabletSharedPtr& tablet = compaction_task->tablet();
     DataDir* data_dir = tablet->data_dir();
-    auto p = _running_tasks.insert(compaction_task);
-    if (!p.second) {
+    bool success;
+    auto iter = _running_tasks.find(tablet->tablet_id());
```
[the remark this answers is not in the document]
Comment (choubenson) [reply be/src/storage/compaction_manager.cpp:336]: Could you please describe it in more detail? You mean that after a task of one tablet is registered in `register_task`  and before it is executed in `run_impl`, while new task of this tablet is picked at the same time, and why is there an inconsistency?

[the remark this answers is not in the document]
Comment (chaoyli) [reply be/src/storage/compaction_manager.cpp:336]: The actual running task in `run_impl` is not the same with the info in `running_tasks`?

[the remark this answers is not in the document]
Comment (choubenson) [reply be/src/storage/compaction_manager.cpp:336]: The actual running task is the same as the info in `running_tasks`.

File: be/src/storage/compaction_manager.cpp [hunk 4]
```diff
@@ -190,16 +190,10 @@ bool CompactionManager::_check_precondition(const CompactionCandidate& candidate
...
-        // tablet already has a running compaction task, skip it
-        VLOG(2) << "skip tablet:" << tablet->tablet_id() << " because there is another running compaction task.";
-        return false;
-    }
-
     int64_t last_failure_ts = 0;
     DataDir* data_dir = tablet->data_dir();
     if (candidate.type == CUMULATIVE_COMPACTION) {
-        std::unique_lock lk(tablet->get_cumulative_lock(), std::try_to_lock);
+        std::shared_lock lk(tablet->get_cumulative_lock(), std::try_to_lock);
```
[the remark this answers is not in the document]
Comment (choubenson) [reply be/src/storage/compaction_manager.cpp:196]: Resolved.

File: be/src/storage/compaction_manager.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (chaoyli) [reply be/src/storage/compaction_manager.cpp:336]: 1. `register_task`

[the remark this answers is not in the document]
Comment (choubenson) [reply be/src/storage/compaction_manager.cpp:336]: Both A and B will in the `running_tasks`. `running_tasks` is a map structure, which is tabletId -> compactionTaskSet

[the remark this answers is not in the document]
Comment (chaoyli) [reply be/src/storage/compaction_manager.cpp:336]: Get it.

**Review by chaoyli — APPROVED** [review]

**Pull request merged**